staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1607 @@
+                        "Failed to create fence: fence_id: {fence_id} fence_response: \
+                         {fence_response}"
+                    );
+                    response = fence_response;
+                }
+            }
+        }
+
+        // Prepare the response now, even if it is going to wait until
+        // fence is complete.
+        let response_len = response
+            .encode(flags, fence_id, ctx_id, ring_idx, writer)
+            .map_err(Error::GpuResponseEncode)?;
+
+        let add_to_queue = if flags & VIRTIO_GPU_FLAG_FENCE != 0 {
+            let ring = match flags & VIRTIO_GPU_FLAG_INFO_RING_IDX {
+                0 => VirtioGpuRing::Global,
+                _ => VirtioGpuRing::ContextSpecific { ctx_id, ring_idx },
+            };
+            debug!("Trying to process_fence for the command");
+            virtio_gpu.process_fence(ring, fence_id, head_index, response_len)
+        } else {
+            true
+        };
+
+        if add_to_queue {
+            vring
+                .add_used(head_index, response_len)
+                .map_err(Error::QueueAddUsed)?;
+            trace!("add_used {}bytes", response_len);
+            *signal_used_queue = true;
+        }
+        Ok(())
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_queue(&self, virtio_gpu: &mut impl VirtioGpu, vring: &VringRwLock) -> Result<()> {
+        let mem = self.mem.as_ref().unwrap().memory().into_inner();
+        let desc_chains: Vec<_> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(mem.clone())
+            .map_err(Error::CreateIteratorDescChain)?
+            .collect();
+
+        let mut signal_used_queue = false;
+        for desc_chain in desc_chains {
+            let head_index = desc_chain.head_index();
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(Error::CreateReader)?;
+            let mut writer = desc_chain.writer(&mem).map_err(Error::CreateWriter)?;
+
+            self.process_queue_chain(
+                virtio_gpu,
+                vring,
+                head_index,
+                &mut reader,
+                &mut writer,
+                &mut signal_used_queue,
+            )?;
+        }
+
+        if signal_used_queue {
+            debug!("Notifying used queue");
+            vring
+                .signal_used_queue()
+                .map_err(Error::NotificationFailed)?;
+        }
+        debug!("Processing control queue finished");
+
+        Ok(())
+    }
+
+    fn handle_event(
+        &self,
+        device_event: u16,
+        virtio_gpu: &mut impl VirtioGpu,
+        vrings: &[VringRwLock],
+    ) -> IoResult<()> {
+        match device_event {
+            CONTROL_QUEUE | CURSOR_QUEUE => {
+                let vring = &vrings
+                    .get(device_event as usize)
+                    .ok_or_else(|| Error::HandleEventUnknown)?;
```

Comment:
Does clippy show a warning if you do:

```suggestion
                    .ok_or(Error::HandleEventUnknown)?;
```

? Out of curiosity.

Reply:
clippy is fine, your suggestion avoids an unnecessary closure. Will change that.